### PR TITLE
Fix muxed conn shared by candidate

### PR DESCRIPTION
**Pull request by cnderrauber**
When we have multiple host candidates and been mux to a single port, if these candidates share a same conn (either tcp or udp), they might read other's msg
cause ice connection cost long time or failed.

This should also fix #473 

#### Reference issue
Fixes #...

### CODE REVIEW

File: udp_mux.go [hunk 1]
```diff
@@ -40,10 +45,22 @@ type UDPMuxDefault struct {
 
 const maxAddrSize = 512
 
+type UdpMuxConn interface {
+	net.PacketConn
+
+	// ReadMsgUdp used to get detination address when received a udp packet
```
Comment (boks1971) [inline udp_mux.go:51]: :nit: destination

**Review by boks1971 — COMMENTED** [review]
Looks good to me but I am not well versed with the depths of this.

**Comment by davidzhao** [comment]
Just tying to understand the change better. Is the situation related to IPv6? If we are looking up by ufrag, it seems that it shouldn't mix up between agents?

**Comment by cnderrauber** [comment]
> Just tying to understand the change better. Is the situation related to IPv6? If we are looking up by ufrag, it seems that it shouldn't mix up between agents?

No, it happens when sfu have multiple host address can be accessed from client. we are not mix up between agents, but candidates, since candidate of single agents share same ufrag.

**Review by davidzhao — APPROVED** [review]
I understand now, thank you for fixing this!

**Comment by davidzhao** [comment]
> No, it happens when sfu have multiple host address can be accessed from client. we are not mix up between agents, but candidates, since candidates of single agent share same ufrag. This issue not caused by IPV6, but easy reproduced when ipv6 enabled.

Another question. If we are using TCP/UDP muxes, why would we add several candidates for different IPs? Since listening is performed by the user, it seems that returning a single muxed candidate per network type would be sufficient.

**Comment by cnderrauber** [comment]
> > No, it happens when sfu have multiple host address can be accessed from client. we are not mix up between agents, but candidates, since candidates of single agent share same ufrag. This issue not caused by IPV6, but easy reproduced when ipv6 enabled.
> 
> Another question. If we are using TCP/UDP muxes, why would we add several candidates for different IPs? Since listening is performed by the user, it seems that returning a single muxed candidate per network type would be sufficient.

we can't return only one candidate since user may listen at all network interfaces (0.0.0.0 and ::).

**Pull request merged**